platform/linux-generic/pktio/mdev/cxgb4.c

```diff
@@ -0,0 +1,887 @@
+
+	cxgb4_wait_link_up(pktio_entry);
+
+	ODP_DBG("%s: open %s is successful\n", MODULE_NAME,
+		pkt_cxgb4->mdev.if_name);
+
+	return 0;
+
+out:
+	cxgb4_close(pktio_entry);
+	return -1;
+}
+
+static int cxgb4_close(pktio_entry_t *pktio_entry)
+{
+	pktio_ops_cxgb4_data_t *pkt_cxgb4 = pktio_entry->s.ops_data;
+
+	ODP_DBG("%s: close %s\n", MODULE_NAME, pkt_cxgb4->mdev.if_name);
+
+	mdev_device_destroy(&pkt_cxgb4->mdev);
+
+	for (uint16_t i = 0; i < pkt_cxgb4->capa.max_input_queues; i++) {
+		cxgb4_rx_queue_t *rxq = &pkt_cxgb4->rx_queues[i];
+
+		if (rxq->rx_data.size)
+			mdev_dma_area_free(&pkt_cxgb4->mdev, &rxq->rx_data);
+	}
+
+	for (uint16_t i = 0; i < pkt_cxgb4->capa.max_output_queues; i++) {
+		cxgb4_tx_queue_t *txq = &pkt_cxgb4->tx_queues[i];
+
+		if (txq->tx_data.size)
+			mdev_dma_area_free(&pkt_cxgb4->mdev, &txq->tx_data);
+	}
+
+	if (pkt_cxgb4->sockfd != -1)
+		close(pkt_cxgb4->sockfd);
+
+	ODP_OPS_DATA_FREE(pkt_cxgb4);
+
+	return 0;
+}
+
+static void cxgb4_rx_refill(cxgb4_rx_queue_t *rxq, uint8_t num)
+{
+	rxq->commit_pending += num;
+
+	while (num) {
+		uint64_t iova = rxq->rx_data.iova + rxq->pidx * ODP_PAGE_SIZE;
+
+		rxq->free_list[rxq->pidx] = odp_cpu_to_be_64(iova);
+
+		rxq->pidx++;
+		if (odp_unlikely(rxq->pidx >= rxq->free_list_len))
+			rxq->pidx = 0;
+
+		num--;
+	}
+
+	/* We commit free list entries to HW in packs of 8 */
+	if (rxq->commit_pending >= 8) {
+		uint32_t val = rxq->doorbell_fl_key | (rxq->commit_pending / 8);
+
+		/* Ring the doorbell */
+		odpdrv_mmio_u32le_write(val, rxq->doorbell_fl);
+
+		rxq->commit_pending &= 7;
+	}
+}
+
+static int cxgb4_recv(pktio_entry_t *pktio_entry,
+		      int rxq_idx, odp_packet_t pkt_table[], int num)
+{
+	pktio_ops_cxgb4_data_t *pkt_cxgb4 = pktio_entry->s.ops_data;
+	cxgb4_rx_queue_t *rxq = &pkt_cxgb4->rx_queues[rxq_idx];
+	uint16_t refill_count = 0;
+	int rx_pkts = 0;
+
+	if (!pkt_cxgb4->lockless_rx)
+		odp_ticketlock_lock(&rxq->lock);
+
+	while (rx_pkts < num) {
+		volatile cxgb4_rx_desc_t *rxd = &rxq->rx_descs[rxq->rx_next];
+		odp_packet_t pkt;
+		uint32_t pkt_len;
+		uint8_t type;
+
+		if (RX_DESC_TO_GEN(rxd) != rxq->gen)
+			break;
+
+		type = RX_DESC_TO_TYPE(rxd);
+
+		if (odp_unlikely(type != RX_DESC_TYPE_FLBUF_X)) {
+			ODP_ERR("Invalid rxd type %u\n", type);
+
+			rxq->rx_next++;
+			if (odp_unlikely(rxq->rx_next >= rxq->rx_queue_len)) {
+				rxq->rx_next = 0;
+				rxq->gen ^= 1;
+			}
+
+			continue;
+		}
+
+		pkt_len = odp_be_to_cpu_32(rxd->pldbuflen_qid);
+
+		/*
+		 * HW skips trailing area in current RX buffer and starts in the
+		 * next one from the beginning.
+		 */
+		if (pkt_len & RX_DESC_NEW_BUF_FLAG) {
+			rxq->cidx++;
+			if (odp_unlikely(rxq->cidx >= rxq->free_list_len))
+				rxq->cidx = 0;
+
+			rxq->offset = 0;
+			refill_count++;
+
+			pkt_len ^= RX_DESC_NEW_BUF_FLAG;
+		}
+
+		if (odp_unlikely(rxq->offset + pkt_len > ODP_PAGE_SIZE)) {
+			/* TODO: reset the HW and reinit ? */
+			ODP_ABORT("Packet write beyond buffer boundary\n");
```

Comment:
I would say that this condition should never ever happen. If it still happens -- it's a fatal error which requires system reboot because it's actually random memory corruption and you can't count on system behaving well after that.

Furthermore, with wonderful Chelsio HW it's impossible to reset it.

We can put cxgb4 pktio into error state permanently if you think it will be enough (with respect to memory corruption statement).

Reply:
Yes, that way the application at least gets the opportunity to do something. Always best to let the application decide if termination is necessary.